### PR TITLE
find: Implement `-samefile`

### PR DESCRIPTION
**Pull request by hanbings**
implement: https://github.com/uutils/findutils/issues/373

BFS test `+1 PASS` and GNU test `PASS +8` / `FAIL -20` in commit bc8ae9859ef41b21a67ab0b72143751240659e3a

### CODE REVIEW

File: tests/find_cmd_tests.rs [hunk 1]
```diff
@@ -587,3 +587,54 @@ fn find_newer_xy() {
         }
     }
 }
+
+#[test]
+#[serial(working_dir)]
+#[cfg(target_os = "linux")]
```
Comment (tavianator) [inline tests/find_cmd_tests.rs:593]:
```suggestion
#[cfg(unix)]
```

Comment (hanbings) [reply tests/find_cmd_tests.rs]: Thanks. Modified in latest commit. This target is more suitable than `linux`. : )

File: src/find/matchers/samefile.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    pub fn new(metadata: Metadata) -> SameFileMatcher {
+        SameFileMatcher { metadata }
+    }
+}
+
+impl Matcher for SameFileMatcher {
+    #[cfg(unix)]
+    fn matches(&self, file_info: &walkdir::DirEntry, _matcher_io: &mut super::MatcherIO) -> bool {
+        let meta = file_info.metadata().unwrap();
+        meta.ino() == self.metadata.ino()
```
Comment (tavianator) [inline src/find/matchers/samefile.rs:26]: You also need to check `meta.dev()`.  Otherwise you'll return true for two unrelated files on different filesystems that happen to get the same inode number.

**Review by tavianator — CHANGES_REQUESTED** [review]

File: src/find/matchers/samefile.rs (hunk 2 above)
Comment (hanbings) [reply src/find/matchers/samefile.rs:31]: Okay, I've added a check.

Comment (sylvestre) [reply src/find/matchers/samefile.rs:31]: we have 
https://github.com/uutils/coreutils/blob/8cac375dddb4d4e87519c137c45fa831ac7f2619/src/uucore/src/lib/features/fs.rs#L559
it would not be a good candidate ?

Comment (hanbings) [reply src/find/matchers/samefile.rs]: > we have https://github.com/uutils/coreutils/blob/8cac375dddb4d4e87519c137c45fa831ac7f2619/src/uucore/src/lib/features/fs.rs#L559 it would not be a good candidate ?

Thanks! This has now been changed to `uucore::fs::paths_refer_to_same_file` in https://github.com/uutils/findutils/pull/389/commits/f612c0da06859beed4aeab498f6195eafe8de7b2 and it works great!

**Comment by sylvestre** [comment]
And I waiting for https://github.com/uutils/findutils/pull/400 to review this first :)

**Comment by sylvestre** [comment]
@hanbings is that expected ? :)
_PASS +8 / FAIL -20 / ERROR +0 / SKIP +0_

**Comment by hanbings** [comment]
If we mean the GNU/BFS tests for the code in this PR itself, yes.

File: src/find/matchers/samefile.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+impl Matcher for SameFileMatcher {
+    #[cfg(unix)]
+    fn matches(&self, file_info: &walkdir::DirEntry, _matcher_io: &mut super::MatcherIO) -> bool {
+        uucore::fs::paths_refer_to_same_file(file_info.path(), &self.path, true)
+    }
+
+    #[cfg(not(unix))]
+    fn matches(&self, _file_info: &walkdir::DirEntry, _matcher_io: &mut super::MatcherIO) -> bool {
+        // FIXME
```
Comment (sylvestre) [inline src/find/matchers/samefile.rs:28]: please create an issue for this once it landed

Comment (sylvestre) [reply src/find/matchers/samefile.rs:28]: also, it supports windows?
https://github.com/uutils/coreutils/blob/92665144c9ddf100d5044dc0c52af122a94587d0/src/uu/split/src/platform/windows.rs#L47

Comment (hanbings) [reply src/find/matchers/samefile.rs]: > also, it supports windows? https://github.com/uutils/coreutils/blob/92665144c9ddf100d5044dc0c52af122a94587d0/src/uu/split/src/platform/windows.rs#L47

Oh! That's right! Changed to support multiple platforms. Thanks!
